utcp.c

```diff
@@ -377,6 +386,61 @@ uint32_t buffer_free(const struct buffer *buf) {
     return buf->maxsize - buf->used;
 }
 
+static void utcp_send_error(const struct pkt_t *pkt, size_t len, ssize_t sent, bool drop) {
+    if(sent != len) {
+        if(sent > len) {
+            debug("Error: sent packet %u and ack %u but with a larger size than it should, %u of %u bytes sent", pkt->hdr.seq, pkt->hdr.ack, sent, len);
+        }
+        else if(sent >= 0) {
+            // we do not handle split packets
+            debug("Warning: failed to send packet %u and ack %u with only %u of %u bytes packet size sent, %s", pkt->hdr.seq, pkt->hdr.ack, sent, len, drop? "dropping the packet" : "retrying it later");
+        }
+        else if(sent == UTCP_WOULDBLOCK) {
+            debug("Debug: failed to send packet %u and ack %u with UTCP_WOULDBLOCK, %s", pkt->hdr.seq, pkt->hdr.ack, drop? "dropping the packet" : "retrying it later");
+        }
+        else {
+            // the pkt receiver might have gone offline causing the routing to fail
+            // drop the packet and continue
+            debug("Error: failed to send packet %u and ack %u with error %u, %s", pkt->hdr.seq, pkt->hdr.ack, sent, drop? "dropping the packet" : "retrying it later");
+        }
+    }
+    else if(drop) {
+        debug("Error: failed to send packet %u and ack %u, dropping the packet [sent=%u]", pkt->hdr.seq, pkt->hdr.ack, sent);
+    }
+}
+
+static bool utcp_send_packet_or_queue(struct utcp *utcp, const struct pkt_t *pkt, size_t len) {
+    ssize_t sent = utcp->send(utcp, pkt, len);
```

Comment:
overworked it again to also prefer to send pending queued packets to response or state change acks plus to remember when at least one packet should be sent next try when no ack could be sent